confdb/aclocal_libs.m4

```diff
@@ -64,11 +64,18 @@ dnl The xxx is specified in the "with_option" parameter.
 dnl
 dnl NOTE: This macro expects a corresponding PAC_SET_HEADER_LIB_PATH
 dnl macro (or equivalent logic) to be used before this macro is used.
+dnl
+dnl NOTE: since setting LIBS may break runtime checks (e.g. AC_CHECK_SIZEOF), we
```

Comment:
To be honest, the way we add directly to CPPFLAGS and LDFLAGS without PUSH/POP kind of bothers me. We end up accumulating a bunch of flags that get used later in the configure script incorrectly, IMO. I wonder if it would be better to instead substitute them as automake flags using for building specific libs or binaries. Thoughts?

Reply:
> Shouldn't the caller PUSH/POP LIBS before and after this macro? 

That's an option. But we need do that at every caller site, versus just here once for all. Same result, right?

> Using WRAPPER_LIBS seems like a hack. That variable is meant to accumulate libraries to be added to MPICH wrapper scripts (mpicc, etc.). I think you point out in a later commit that it is too MPICH specific.

Yes, I agree. However, it will need a separate PR to just clean out the current usage of `WRAPPER_LIBS`. Potentially, we'll use another variable, e.g. `EXTERN_LIBS`, for building mpich, and use `WRAPPER_LIBS` only for wrapper scripts. `WRAPPER_LIBS` can easily include `EXTERN_LIBS` at the end if that is appropriate.

Reply:
> To be honest, the way we add directly to CPPFLAGS and LDFLAGS without PUSH/POP kind of bothers me. We end up accumulating a bunch of flags that get used later in the configure script incorrectly, IMO. I wonder if it would be better to instead substitute them as automake flags using for building specific libs or binaries. Thoughts?

To directly modify these precious flags and have them affect how conftests behave is, like it or not, the design of `autoconf`. Sometimes we need these flags modified as we configure as later tests may depend on them. Certainly a "cleaner" approach is to understand on a per-package basis and add extra-variables as you said. But that is extra code and logic and adds to the maintenance.

Reply:
> > To be honest, the way we add directly to CPPFLAGS and LDFLAGS without PUSH/POP kind of bothers me. We end up accumulating a bunch of flags that get used later in the configure script incorrectly, IMO. I wonder if it would be better to instead substitute them as automake flags using for building specific libs or binaries. Thoughts?
> 
> To directly modify these precious flags and have them affect how conftests behave is, like it or not, the design of `autoconf`. Sometimes we need these flags modified as we configure as later tests may depend on them. Certainly a "cleaner" approach is to understand on a per-package basis and add extra-variables as you said. But that is extra code and logic and adds to the maintenance.

Hmm, I suppose you are right about the other flags. I think my main annoyance is actually that they persist all the way to the wrapper scripts even when they are not needed.

Back to the `WRAPPER_LIBS` issue, I'm surprised to see that `WRAPPER_LIBS` actually exists in configure scripts that are not just the main MPICH one. It pains to think of introducing it to `confdb`, but perhaps it is a decent half-step.

Reply:
> > Shouldn't the caller PUSH/POP LIBS before and after this macro?
> 
> That's an option. But we need do that at every caller site, versus just here once for all. Same result, right?

All the call sites that are currently doing PUSH/POP (all of them?) are now doing it to no effect, which is confusing to someone reading the script. I've gone back and forth, but I think introducing WRAPPER_LIBS into confdb is a mistake.

Reply:
Does it make a difference if we replace `WRAPPER_LIB` with `EXTERNAL_LIB`? Of course all the extra push/pop should get removed at some point. It's not a lot of place, right? currently we only have limited external libraries -- handful places, maybe? 😁

Reply:
I guess for me it comes down to more deeply ingraining `WRAPPER_LIBS` into confdb and non-MPICH build files vs. using a more clear and representative name. None of this has to do specifically with hwloc, however. Is there a way we can fix the user issue without having to resolve this first?

Reply:
I see. Your PR #3540 works in term of external linking with hwloc, right? I won't mind merge that PR first. 

That said, I don't think we need a push for a cursory job for this PR.

The motivation for this PR is to bring more consistency in the way we configure external libraries -- one of which is to use `PAC_CHECK_HEADER_LIB`. That macro always had issue of modifying `LIBS` resulting config test failures. We have been sporadically patching it for individual usage, again, brings inconsistency. The solution, IMO, is to "stage" LIBS rather than modify it directly. I used `WRAPPER_LIBS` because we had been using it (in those sporadic patches I mentioned). That is a misuse as "WRAPPER" had an original specific meaning. So I would like to have a separate name for it. Are we on the same page?

Reply:
Yes, we are on the same page. Should I update my original PR in the interim?

Reply:
Sure.